### ISSUE TITLE
perf(solve): stop recomputing the residual in the Newton forward loop (line-search early exit + carried residual)

### ISSUE THREAD

**Issue by ctdegroot** [DeGrootResearchGroup/aquaflux#108]
Surfaced by a code-wide speed review. Two redundant-residual problems in the same loop; natural to fix together.

## 1. The default line search evaluates every rung unconditionally

`aquaflux/solve/implicit.py:102-110`. The backtracking ladder is a Python `for` loop, so all `line_search_steps + 1` rungs are unrolled into the graph and **every** rung's `residual_fn(phi + alpha*delta)` is evaluated; `jnp.where` / `accepted` only *select* afterwards.

Measured, one step on a scalar residual:

```
line_search=10 -> 13 residual evaluations
line_search=0  ->  2
```

With the default `line_search=10` that is ~11 extra full residual assemblies per Newton step even when alpha=1 is accepted at the first rung — which is the case near the root and for **every** step of a linear or affine residual. It also multiplies the traced graph ~11x, a direct compile-time cost.

Note the current comment at :100 justifies the unrolled form as keeping the step "a constant-length operation, so it composes with the `while_loop` and the IFT `custom_vjp` without a data-dependent branch." A `lax.while_loop` over the ladder satisfies that constraint equally well — it is a bounded loop with no data-dependent *shape* — and the repo already has this exact pattern one file over in `PseudoTransientStep.stepper` (`aquaflux/solve/continuation.py:278-291`). The inconsistency looks historical.

Caveat on magnitude: against a 67-127-iteration outer GMRES the runtime share is ~5%. The graph-size and compile share is larger, and the fix is strictly free.

## 2. The residual at each iterate is computed twice per Newton iteration

`aquaflux/solve/implicit.py:196-203`: `_forward`'s `body` ends with `jnp.linalg.norm(residual_fn(phi, theta))` for the loop condition. The next iteration's `forward_step_fn` then recomputes `residual_fn(phi)` from scratch — `_damped_newton_step` -> `newton_correction` (line 72), and `PseudoTransientStep.step` (`aquaflux/solve/continuation.py:240`). One redundant full residual assembly per Newton iteration.

The continuation path is worse: `attempt` (`continuation.py:269`) already computes `jnp.linalg.norm(residual_fn(candidate))`, which is *exactly* the quantity `_forward`'s body then recomputes on the same field.

`newton_correction` already returns `(delta, r)` for precisely this reason — its docstring says so ("so a caller doing a line search need not recompute it"). The plumbing just stops one level short of the loop carry.

## Proposed fix

- Replace the unrolled ladder with a `lax.while_loop` that exits on the first accepted rung.
- Widen the `_forward` loop carry to `(phi, r, norm, step)` and have `ForwardStep` return `(phi_next, r_next)` so the condition norm and the next step's residual are the same evaluation.

Both touch the `ForwardStep` protocol, so doing them in one change avoids widening the interface twice.

## Acceptance

- A linear residual converges with 2 residual evaluations per Newton step, not 13.
- One residual assembly per Newton iteration, not two.
- Existing tests unchanged: `test_implicit_solve.py` gradient-matches-closed-form to 1e-10 and stays iteration-count-independent; the open-flow cases that motivated the line search still converge.

**Comment by ctdegroot** [comment]
**Part 1 is done; part 2 is not. Keeping this open for part 2 only.**

Checked against `main` (`85b83ec`):

**1. Unrolled ladder — FIXED.** `backtracking_line_search` is now a `jax.lax.while_loop` that stops at the first (largest) admissible rung (`solve/implicit.py`, the `while_loop` at ~:428), so a step whose full length already descends costs **one** residual evaluation rather than `steps + 1`. The body compiles once instead of unrolling `steps + 1` copies into the graph, which also removes the compile-time cost noted here. The acceptance criterion "a linear residual converges with 2 residual evaluations per Newton step, not 13" is met by construction.

Two things the fix picked up that were not in this issue's scope and are worth knowing before touching it: the ladder now also extends *above* `alpha = 1` (`grow` rungs), and the no-admissible-rung fallback is the longest **finite** rung no longer than the full step — not the shortest, which was a guaranteed stall.

**2. Residual computed twice per Newton iteration — NOT done.** `_forward`'s carry is still `(phi, step, residual_norm)`, and its body still ends with `norm_fn(residual_fn(phi, theta))` while the next iteration's step recomputes the residual from scratch. The proposed fix — widen the carry and have the step return `(phi_next, r_next)` — is unbuilt.

Note the remaining half now costs more to do than when this was filed: `StepOutcome` has grown to seven fields, so returning the residual means an eighth, and the `ForwardStep` contract now lives in its own module (`solve/forward_step.py`) with two implementations plus the dual-time step. Still worth it — one full residual assembly per Newton iteration — but it is no longer a small change, and the "do both together to avoid widening the interface twice" argument no longer applies, since part 1 landed without touching the protocol.
